### PR TITLE
Refactored Networking Layer

### PR DESCRIPTION
**Pull request by Mintri1199**
TLDR: 
* Opt for protocol oriented networking layer

Currently our networking layer does basic GET request to call the Unsplash API for searching pictures with a given keyword. However, the configuration for networking methods is very hard coded and doesn't have room for code reusability. Another consideration that drive the decision to refactor is that it is not very testable. As the project gets bigger with more methods and functionalities (authentication) it is important that this app should have a networking layer that is modular while be able to test its components in isolation. 

This PR aims to introduce a protocol oriented approach to networking to enhance it testability, code reusability, error messages, and type safety. The majority of this code was inspired from a [medium article] in 2018 where the author demonstrates isolating components that would involved in simple/complex methods by using a protocol oriented approach. 

The author template work as follow:
* An `Endpoint` protocol that has baseURL, path, httpMethod, tasks, and headers attributes.
* Create an `API Endpoint` public enum that would conform to `Endpoint` protocol. This allows us to easily configure networking components based on what actions we set in our case. 
* Create a generic `Router` class that would build request and configure the headers and parameters based on what `Endpoint` type that was passed in.
* The `NetworkManager` class will have router attribute that is specific to whatever API that have been set.

The benefit that is apparent when using this approach is code separation. It moved the URL builder, URLSession, and endpoint configurations out of the main `NetworkManager` class. This help clearly define the intentions of each components and API endpoints while avoiding massive `Manager` class.

In the past I also improved the author's template by using `Result` type when writing methods in the `NetworkManager` class for clear and type safe asynchronous code. 

The refactored network layer and its functionalities are integrated to the `SearchViewController`. The next steps from this is to figure out how to tests each components in isolation and how to test the networking methods before trying to implement a CI/CD system.

[medium article]: https://medium.com/flawless-app-stories/writing-network-layer-in-swift-protocol-oriented-approach-4fa40ef1f908

### CODE REVIEW

File: WallPaperDev/WallPaperDev/Resources/en.lproj/Localizable.strings [hunk 1]
```diff
@@ -45,7 +45,7 @@
 "message_for_goal_with_no_description" = "There is no description for this goal";
 "view_action" = "View";
 
-"tutorial_title_1" = "Welcome to Kamigami";
+"tutorial_title_1" = "Welcome to Kabegami";
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Resources/en.lproj/Localizable.strings:48]: Like the cat? :)

File: WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCell.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+//  WallPaperDev
+//
+//  Created by Jamar Gibbs on 11/5/19.
+//  Copyright © 2019 Stephen Ouyang. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+class SearchImagesCell : UICollectionViewCell {
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCell.swift:12]:
```suggestion
final class SearchImagesCell: UICollectionViewCell {
```

File: WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCell.swift [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  let identifier = "searchCell"
+  
+  // Using layers to hold uninteractable images
+  lazy var photoLayer = CALayer()
+  lazy var maskLayer = CAShapeLayer()
+  lazy var borderLayer = CAShapeLayer()
+
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCell.swift:19]: Are you noticing any flickering for the cells? May make sense to when a cell is preparing to be reused we empty out the old CALayer? Thoughts?

File: WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCVLayout.swift [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+//  SearchImagesCVLayout.swift
+//  WallPaperDev
+//
+//  Created by Jackson Ho on 9/16/19.
+//  Copyright © 2019 Stephen Ouyang. All rights reserved.
+//
+
+import UIKit
+
+class SearchImagesCVLayout: UICollectionViewFlowLayout {
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCVLayout.swift:11]:
```suggestion
final class SearchImagesCVLayout: UICollectionViewFlowLayout {
```

File: WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCV.swift [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+//  SearchImagesCV.swift
+//  WallPaperDev
+//
+//  Created by Jamar Gibbs on 11/5/19.
+//  Copyright © 2019 Stephen Ouyang. All rights reserved.
+//
+
+import UIKit
+
+class SearchImagesCV : UICollectionView {
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCV.swift:11]: In general if we are not subclassing a class let's mark it as final

File: WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImageCell.swift [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+//  SearchImagesCell.swift
+//  WallPaperDev
+//
+//  Created by Jamar Gibbs on 12/3/19.
+//  Copyright © 2019 Stephen Ouyang. All rights reserved.
+//
+
+import UIKit
+
+class SearchImagesCell : UICollectionViewCell {
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImageCell.swift:11]: Appears we have two `class SearchImagesCell : UICollectionViewCell {`?

File: WallPaperDev/WallPaperDev/Views/CreateImage Views/ImageSelectionCell.swift [hunk 7]
```diff
@@ -74,40 +54,34 @@ extension ImageSelectionCell {
...
-    arrowLayer.fillColor = UIColor.clear.cgColor
-    arrowLayer.lineJoin = CAShapeLayerLineJoin.round
-    arrowLayer.lineCap = CAShapeLayerLineCap.round
-    showMoreButton.layer.addSublayer(arrowLayer)
+    let sizeConstant: CGFloat = bounds.width / 3
+    showMoreLayer.frame = CGRect(origin: bounds.origin, size: CGSize(width: sizeConstant, height: sizeConstant))
+    showMoreLayer.anchorPoint = CGPoint(x: 0.5, y: 0.5)
+    showMoreLayer.position = CGPoint(x: bounds.midX, y: bounds.midY + showMoreLayer.frame.height)
+    showMoreLayer.cornerRadius = sizeConstant / 2
+    showMoreLayer.backgroundColor = UIColor.lightGray.cgColor
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/CreateImage Views/ImageSelectionCell.swift:72]: Thoughts on either using theme or adding a light gray to the theme?

File: WallPaperDev/WallPaperDev/Views/CreateImage Views/ImageSelectionCell.swift [hunk 8]
```diff
@@ -74,40 +54,34 @@ extension ImageSelectionCell {
...
+
+    let arrowPath = UIBezierPath.arrow(from: CGPoint(x: sizeConstant / 4, y: sizeConstant / 2),
+                                       to: CGPoint(x: sizeConstant * 0.75, y: sizeConstant / 2),
+                                       tailWidth: sizeConstant / 15, headWidth: sizeConstant / 4,
+                                       headLength: sizeConstant / 6)
+
+    showMoreLayer.strokeColor = UIColor.white.cgColor
+    showMoreLayer.lineWidth = 3
+    showMoreLayer.path = arrowPath.cgPath
+    showMoreLayer.fillColor = UIColor.white.cgColor
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/CreateImage Views/ImageSelectionCell.swift:82]: Similar comments here and above

File: WallPaperDev/WallPaperDev/Views/Controllers/SearchImageViewController.swift [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+  }
+}
+// MARK: - SearchBar Delegate
+extension SearchImageViewController : UISearchBarDelegate {
+  func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
+    // set urls to empty for new search
+    viewModel.imageURLS = []
+    let keywords = searchBar.text
+    guard let finalKeyword = keywords?.replacingOccurrences(of: " ", with: "+") else { return }
+    viewModel.loadSearchURLS(for: finalKeyword, cv: searchImagesCV)
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/Controllers/SearchImageViewController.swift:141]: Let's consider passing the raw search term to the view model and then have it do the replacements

File: WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCell.swift (hunk 3 above)
Comment (Mintri1199) [reply WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCell.swift:19]: Oh yeah I just see it now. I'll issue a bug to work on this later.

File: WallPaperDev/WallPaperDev/Resources/en.lproj/Localizable.strings (hunk 1 above)
Comment (Mintri1199) [reply WallPaperDev/WallPaperDev/Resources/en.lproj/Localizable.strings:48]: Oh I didn't even realized that. 😆

File: WallPaperDev/WallPaperDev/Models/Photos.swift [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+  let total, totalPages: Int
+  let results: [PhotosUrls]
+}
+
+struct PhotosUrls: Codable {
+  let urls: Urls
+}
+
+struct Urls: Codable {
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Models/Photos.swift:26]: We should want to consider prefacing these structs: `Urls` and `PhotoUrls` with `Unsplash`

File: WallPaperDev/WallPaperDev/Networking/NetworkServiceProtocol.swift [hunk 11]
```diff
@@ -0,0 +1,12 @@
+//
+// Created by Jackson Ho on 3/17/20.
+// Copyright (c) 2020 Stephen Ouyang. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+protocol NetworkServiceProtocol {
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Networking/NetworkServiceProtocol.swift:9]: Thoughts on renaming to `UnsplashAPIService`

Comment (alexander-dejeu) [reply WallPaperDev/WallPaperDev/Networking/NetworkServiceProtocol.swift:9]: The rational being that we may have different protocols for each service instead of a single network service

File: WallPaperDev/WallPaperDev/Networking/Reach.swift [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+  var description: String {
+    switch self {
+    case .wwan: return "WWAN"
+    case .wiFi: return "WiFi"
+    }
+  }
+}
+
+enum ReachabilityStatus: CustomStringConvertible {
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Networking/Reach.swift:24]: Nice

File: WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCVLayout.swift [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  override func prepare() {
+    super.prepare()
+    guard let cv = collectionView else {
+      return            
+    }
+    scrollDirection = .vertical
+    let availableWidth = cv.bounds.inset(by: cv.layoutMargins).size.width
+    let cellWidth = Int(availableWidth / 2.3)
+//        itemSize = CGSize(width: cellWidth, height: Int(cv.bounds.height) - 10)
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCVLayout.swift:20]:
```suggestion

```

File: WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImageCell.swift [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+  lazy var borderLayer = CAShapeLayer()
+  
+  var cellImage: UIImage? {
+    didSet {
+      photoLayer.contents = cellImage?.cgImage
+    }
+  }
+  
+  func getImage(_ image: UIImage?) {
+    photoLayer.backgroundColor = UIColor.red.cgColor
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImageCell.swift:26]: Theme

File: WallPaperDev/WallPaperDev/Views/Controllers/SearchImageViewController.swift [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+    ])
+  }
+  
+  private func setupNavBar() {
+    searchController.searchBar.delegate = self
+    searchController.searchBar.placeholder = "Search for images"
+    searchController.searchBar.autocorrectionType = .default
+    searchController.searchBar.sizeToFit()
+    
+    navigationItem.title = "Search for a new image"
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/Controllers/SearchImageViewController.swift:67]: Let's toss these into the localizable file

File: WallPaperDev/WallPaperDev/Views/Controllers/SearchImageViewController.swift [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+      searchImagesCV.leadingAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.leadingAnchor),
+      searchImagesCV.trailingAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.trailingAnchor),
+      searchImagesCV.topAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.topAnchor),
+      searchImagesCV.bottomAnchor.constraint(equalTo: self.view.bottomAnchor)
+    ])
+  }
+  
+  private func setupNavBar() {
+    searchController.searchBar.delegate = self
+    searchController.searchBar.placeholder = "Search for images"
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/Controllers/SearchImageViewController.swift:63]: Similar comment as below.

File: WallPaperDev/WallPaperDev/Networking/NetworkingService.swift [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    
+    guard let url = newURL?.url else {
+      #if DEBUG
+        print("invalid url")
+      #endif
+      return
+    }
+    
+    var request = URLRequest(url: url)
+//        request.addValue("Bearer \(APIKey.key)", forHTTPHeaderField: "Authorization")
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Networking/NetworkingService.swift:37]: By not having an auth token what are the risks right now?

File: WallPaperDev/WallPaperDev/Views/Controllers/SearchImageViewController.swift [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+  }
+
+  func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+    guard let cell = searchImagesCV.dequeueReusableCell(withReuseIdentifier: searchImagesCV.cellID, for: indexPath) as? SearchImagesCell else {
+      return UICollectionViewCell()
+    }
+
+    let urlString = viewModel.imageURLS[indexPath.row].urls.regular
+
+    viewModel.networkManager.getPhotoData(from: urlString) { result in
```
Comment (alexander-dejeu) [inline WallPaperDev/WallPaperDev/Views/Controllers/SearchImageViewController.swift:117]: Not a huge concern right now, but as cells are being recycled we will not kill these previous tasks therefore we will be wasting some resources and experience flickering

**Review by alexander-dejeu — APPROVED** [review]
Incredible work! At the core looks like a great start for the networking layer + feature!

File: WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCVLayout.swift (hunk 13 above)
Comment (Mintri1199) [reply WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImagesCVLayout.swift:20]: The Unsplash Search Screen is currently in process of redesign, so the layout for the next PR will be different.

File: WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImageCell.swift (hunk 14 above)
Comment (Mintri1199) [reply WallPaperDev/WallPaperDev/Views/SearchImages Views/SearchImageCell.swift:26]: This was just a placeholder color to show the cell. I'll remove it soon.

File: WallPaperDev/WallPaperDev/Views/Controllers/SearchImageViewController.swift (hunk 15 above)
Comment (Mintri1199) [reply WallPaperDev/WallPaperDev/Views/Controllers/SearchImageViewController.swift:67]: I think it will look better if I remove the title since it reduce the number of elements in the nav bar. Thus avoid the cramped look.

File: WallPaperDev/WallPaperDev/Networking/NetworkingService.swift (hunk 17 above)
Comment (Mintri1199) [reply WallPaperDev/WallPaperDev/Networking/NetworkingService.swift:37]: I am removing this file in favor of the new network layer.

File: WallPaperDev/WallPaperDev/Networking/NetworkServiceProtocol.swift (hunk 11 above)
Comment (Mintri1199) [reply WallPaperDev/WallPaperDev/Networking/NetworkServiceProtocol.swift:9]: I think I'm going to move this protocol to the `UnsplashEndpoint` file
